extern/osg-ffmpeg-videoplayer/audiodecoder.cpp

```diff
@@ -159,31 +160,48 @@ int MovieAudioDecoder::audio_decode_frame(AVFrame *frame, int &sample_skip)
 
     for(;;)
     {
-        while(pkt->size > 0)
-        {
-            int len1, got_frame;
-
-            len1 = avcodec_decode_audio4(mAudioContext, frame, &got_frame, pkt);
-            if(len1 < 0) break;
+        /* send the packet with the compressed data to the decoder */
+        int ret = 0;
+        if (mGetNextPacket)
+            ret = avcodec_send_packet(mAudioContext, pkt);
 
-            if(len1 <= pkt->size)
+        /* read all the output frames (in general there may be any number of them */
+        while (ret >= 0)
+        {
+            ret = avcodec_receive_frame(mAudioContext, frame);
+            if (ret == AVERROR_EOF)
             {
-                /* Move the unread data to the front and clear the end bits */
-                int remaining = pkt->size - len1;
-                memmove(pkt->data, &pkt->data[len1], remaining);
-                av_shrink_packet(pkt, remaining);
+                // We decoded all frames in the package.
+                // Flush buffers and decode a next package.
+                mGetNextPacket = true;
+                avcodec_flush_buffers(mAudioContext);
+                continue;
+            }
+            else if (ret == AVERROR(EAGAIN))
+            {
+                // Additional frames were not decoded yet.
+                // Enter to the draining mode.
+                avcodec_send_packet(mAudioContext, NULL);
+                ret = 0;
+                continue;
+            }
```

Comment:
This doesn't look right/logical. `EAGAIN` from `avcodec_receive_frame` means another packet needs to be sent before a frame can be decoded. Sending a NULL packet means the stream is over, so you get an `EOF` that's supposed to tell you to stop. It looks like you're abusing the stream-is-done message to signal to get another packet. `avcodec_flush_buffers` also flushes and restarts decoding, which would cause partial frames, that haven't been fully decoded yet, to drop.

More generally, it doesn't make sense for the decoder thread to be sending the encoded packets and receiving the decoded frames. For example, [note here](https://github.com/kcat/openal-soft/blob/master/examples/alffplay.cpp#L1320) how it holds in a loop while getting `EAGAIN`, signalling to the packet pump thread to send more packets and waiting for it to do so. When it gets a frame, it signals the pump thread to go and send another while it does its thing. When it gets `EOF`, the stream is over and it ends. Then [here](https://github.com/kcat/openal-soft/blob/master/examples/alffplay.cpp#L1523) the packet pump thread gets a packet from the input file, queues it, then sends whatever packets it can to the available stream handlers. If the combined queues hold enough compressed data, it simply waits until it can get rid of some before reading more (to avoid reading the whole file into memory).

This setup improves multithreading performance since it splits the workload across threads. With the decoder thread doing the sending and receiving, you're serializing the workload:
```
send_packet           receive_frame....send_packet           receive_frame....
          |           ^                          |           ^
          v           |                          v           |
          decode packet                          decode packet
```
Instead, you want to be able to interleave the work:
```
send_packet .. send_packet
          |              |
          v              v
          decode packet  decode packet
                      |                |
                      v                v
                      receive_frame .. receive_frame ..
```
That helps makes sure everything stays busy, avoiding wasted cycles while waiting for another thread to complete what you just told it to do.